tests/test_streaming.py

```diff
@@ -175,6 +177,82 @@ def test_dataloader_per_stream_batching(local_remote_dir: Tuple[str, str], batch
     assert batches_seen == total_batches
 
 
+@pytest.mark.parametrize('batch_size', [4, 7])
+@pytest.mark.parametrize('seed', [2222])
+@pytest.mark.parametrize('shuffle', [True])
+@pytest.mark.parametrize('physical_nodes', [2, 8])
+@pytest.mark.parametrize('ranks_per_node', [4, 8])
+@pytest.mark.parametrize('workers_per_rank', [4, 8])
+@pytest.mark.parametrize('num_canonical_nodes', [8])
+@pytest.mark.usefixtures('local_remote_dir')
+def test_dataloader_device_per_stream_batching(local_remote_dir: Tuple[str, str], batch_size: int,
+                                               seed: int, shuffle: bool, physical_nodes: int,
+                                               ranks_per_node: int, workers_per_rank: int,
+                                               num_canonical_nodes: int):
+    # create mock datasets for 2 streams. Second one has 1.5x the samples
+    local, remote = local_remote_dir
+    local1 = os.path.join(local, 'stream1')
+    local2 = os.path.join(local, 'stream2')
+    remote1 = os.path.join(remote, 'stream1')
+    remote2 = os.path.join(remote, 'stream2')
+
+    # stream 1 has samples 0->600, sample ids 0->200
+    convert_to_mds(out_root=remote1,
+                   dataset_name='sequencedataset',
+                   num_samples=200,
+                   size_limit=1 << 8)
+    # stream 2 has samples 600 and above, sample ids 200 and above.
+    # This lets us differentiate between the samples from each stream
+    convert_to_mds(out_root=remote2,
+                   dataset_name='sequencedataset',
+                   num_samples=300,
+                   offset=600,
+                   size_limit=1 << 8)
+
+    stream1 = Stream(local=local1, remote=remote1)
+    stream2 = Stream(local=local2, remote=remote2)
+
+    # Build StreamingDataset
+    dataset = StreamingDataset(streams=[stream1, stream2],
+                               shuffle=shuffle,
+                               batch_size=batch_size,
+                               shuffle_seed=seed,
+                               num_canonical_nodes=num_canonical_nodes,
+                               batching_method='per_stream',
+                               shuffle_block_size=100)
```

Comment:
Can this test be rewritten such that the new batching  is applied to StreamingDataset directly?

Reply:
No because `StreamingDataset` will generate the sample partition only at `__iter__` time, and because these are CPU tests, the `World` object will think that we only have 1 device and 1 node. We want to check the per-device-batch behavior, so this requires us to create a `fake_world` object with multiple nodes and devices.